### DISCUSSION TITLE
Issue with installing "pysr" in Python3 Jupyter Notebook ipynb file (exit status 1)

### DISCUSSION THREAD

**Discussion by unary-code** [MilesCranmer/PySR#357] [Q&A] [answered]
Hi,
I am running Jupyter Notebook on my MacBook computer from a standard computer terminal.
I installed both juliaup and the Julia application itself.
When I type "julia" in the standard terminal on my MacBook, I get this message: "Error: The Julia launcher failed to figure out which juliaup channel to use." Perhaps I should fix this and perhaps this is what is causing "!python3 -m pysr install" command in a Python3 ipynb file to give a error message.

I went to the localhost:[PORT] website to access the Jupyter Notebook interface.
Then I created a Python3 ipynb file.
Then in the cell blocks of this ipynb file (not on a terminal), I ran:
```
!curl -fsSL https://install.julialang.org | sh
```
That code gave me this response: 
```
info: downloading installer
Welcome to Julia!

It seems that Juliaup is already installed on this system. Please remove the previous installation of Juliaup before you try to install a new version.
```

Then I ran 
```
!pip3 install -U pysr
```
This gave me a bunch of "Requirement satisfied" messages. A truncated version of the output is:
```
Requirement already satisfied: pysr in /opt/anaconda3/lib/python3.8/site-packages (0.14.2)
Requirement already satisfied: sympy in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (1.8)
Requirement already satisfied: pandas in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (1.2.4)
Requirement already satisfied: numpy in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (1.23.5)
Requirement already satisfied: scikit-learn>=1.0.0 in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (1.2.2)
Requirement already satisfied: julia>=0.6.0 in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (0.6.1)
Requirement already satisfied: click>=7.0.0 in /opt/anaconda3/lib/python3.8/site-packages (from pysr) (7.1.2)
... (there was more "Requirements satisfieds" but I cut it off here)
```

Then I ran "!python3 -m pysr install", which gave me my error that I am asking about in this post.
That command gave me this output:
```
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/__main__.py", line 4, in <module>
    _cli(prog_name="pysr")
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/_cli/main.py", line 37, in _install
    install(julia_project, quiet, precompile)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/julia_helpers.py", line 76, in install
    _julia_version_assertion()
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/julia_helpers.py", line 270, in _julia_version_assertion
    if not is_julia_version_greater_eq(version=(1, 6, 0)):
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/julia_helpers.py", line 134, in is_julia_version_greater_eq
    juliainfo = _load_juliainfo()
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/julia_helpers.py", line 25, in _load_juliainfo
    juliainfo = JuliaInfo.load(julia="julia")
  File "/opt/anaconda3/lib/python3.8/site-packages/julia/juliainfo.py", line 87, in load
    raise subprocess.CalledProcessError(
subprocess.CalledProcessError: Command '['julia', '-e', '...']' returned non-zero exit status 1.
```

Additional question: In order to use your PySR package, I should be using a Python3 ipynb file, not a Julia file, right?
Because the ultimate code that I write myself to use your PySR package is in Python, not Julia? What I mean, is that I create a PySRRegressor object usign Python code 

```
model = PySRRegressor(
    niterations=40,  # < Increase me for better results
    binary_operators=["+", "*"])
```
is Python not Julia), not using Julia code, therefore I believe I should be using a Python3 ipynb file to run the commands to install PySR.

**Comment by MilesCranmer** [comment]
Hi @unary-code,

In future questions could you wrap your code outputs with three backquotes: `\``? Otherwise it's hard to read your question. I edited your post to add this but just an FYI for future posts :+1:

>  I should be using a Python3 ipynb file, not a Julia file, right?

Correct. Unless you use [SymbolicRegression.jl](github.com/MilesCranmer/SymbolicRegression.jl/) directly in which case you would use the julia kernel.

In general I would recommend using IPython instead of Jupyter notebooks though, as the output is cleaner (it's difficult to get Julia to print nice outputs to Jupyter through Python, so I can't use the regular progress bar)

> Perhaps I should fix this and perhaps this is what is causing "!python3 -m pysr install" command in a Python3 ipynb file to give a error message.

Yes, the error `subprocess.CalledProcessError: Command '['julia', '-e', '...']' returned non-zero exit status 1.` just means it could not start Julia. So you will need to fix Julia first before trying to install PySR. Maybe you could try to remove all of your Julia installations completely and install it again? Be sure to only use juliaup, instead of using a combination of different julia versions (e.g., there is one with conda, and another one you can install from a .dmg, and yet another one you can install with brew).

Cheers,
Miles

**Comment by unary-code** [comment]
Hi,

Thanks, I'll use 3 backslash quotation marks ``` in the future on Github posts.

When I do the following commands, I get this output (does not change between when I have both Julia application and juliaup installed on my computer, and when I only have juliaup installed on my computer):

Do you have an idea of what causes the following errors?

When I did "!python3 -m pysr install", I got a sequence of a lot of lines, but the most important outputted lines were:
```
(some beginning lines of output)
[ Info: Julia executable: /Users/davidkatz/.julia/juliaup/julia-1.9.1+0.aarch64.apple.darwin14/bin/julia
[ Info: Trying to import PyCall...
┌ Error: `import PyCall` failed
│   exception =
│    ArgumentError: Package PyCall [438e738f-606a-5dbb-bf0a-cddfbfd45ab0] is required but does not seem to be installed:
│     - Run `Pkg.instantiate()` to install all recorded dependencies.
│    
│    Stacktrace:
│     [1] _require(pkg::Base.PkgId, env::Nothing)
│       @ Base ./loading.jl:1774
│     [2] _require_prelocked(uuidkey::Base.PkgId, env::Nothing)
│       @ Base ./loading.jl:1660
│     [3] _require_prelocked
│       @ ./loading.jl:1658 [inlined]
│     [4] macro expansion
│       @ ./lock.jl:267 [inlined]
│     [5] require(uuidkey::Base.PkgId)
│       @ Base ./loading.jl:1655
│     [6] top-level scope
│       @ /opt/anaconda3/lib/python3.8/site-packages/julia/install.jl:36
│     [7] include(mod::Module, _path::String)
│       @ Base ./Base.jl:457
│     [8] exec_options(opts::Base.JLOptions)
│       @ Base ./client.jl:307
│     [9] _start()
│       @ Base ./client.jl:522
└ @ Main /opt/anaconda3/lib/python3.8/site-packages/julia/install.jl:38
[ Info: Installing PyCall...
    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
   Installed PyCall ───── v1.96.0
   Installed MacroTools ─ v0.5.10     ]  1/2
mProgress [========================================>]  2/2    Updating `~/.julia/environments/pysr-0.14.2/Project.toml`
  [438e738f] + PyCall v1.96.0
    Updating `~/.julia/environments/pysr-0.14.2/Manifest.toml`
  [8f4d0f93] + Conda v1.9.0
  [682c06a0] + JSON v0.21.4
.... (some other lines)

Building PyCall → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/f33ca06633c2cf08753d235208fcb46a052072ea/build.log`
... (some other lines)

Progress [===========================>             ]  2/3
  ◐ PyCall
  Progress [===========================>             ]  2/3
  ◓ PyCall
  Progress [========================================>]  3/3
  3 dependencies successfully precompiled in 8 seconds. 9 already precompiled.
]0;Julia
Precompiling PyCall...
ImportError: No module named site
]0;JuliaTraceback (most recent call last):
  File "/opt/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/__main__.py", line 4, in <module>
    _cli(prog_name="pysr")
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/_cli/main.py", line 37, in _install
    install(julia_project, quiet, precompile)
  File "/opt/anaconda3/lib/python3.8/site-packages/pysr/julia_helpers.py", line 84, in install
    julia.install(quiet=quiet)
  File "/opt/anaconda3/lib/python3.8/site-packages/julia/tools.py", line 127, in install
    raise PyCallInstallError("Precompiling")
julia.tools.PyCallInstallError: Precompiling PyCall failed.

** Important information from Julia may be printed before Python's Traceback **

Some useful information may also be stored in the build log file
`~/.julia/packages/PyCall/*/deps/build.log`.

```

I was able to do "from pysr import PySRRegressor", but then when I did "model.fit(X, y)", I got this error:
```
Compiling Julia backend...
(some outputted lines)
OSError: dlopen(/Users/[MYUSER]/.julia/juliaup/julia-1.9.1+0.aarch64.apple.darwin14/lib/libjulia.1.9.dylib, 0x000A): tried: '/Users/[MYUSER]/.julia/juliaup/julia-1.9.1+0.aarch64.apple.darwin14/lib/libjulia.1.9.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/local/lib/libjulia.1.9.dylib' (no such file), '/usr/lib/libjulia.1.9.dylib' (no such file)
```

**Comment by MilesCranmer** [comment]
Thanks, this is helpful. The following line indicates the issue:

```
OSError: dlopen(/Users/[MYUSER]/.julia/juliaup/julia-1.9.1+0.aarch64.apple.darwin14/lib/libjulia.1.9.dylib, 0x000A): tried: '/Users/[MYUSER]/.julia/juliaup/julia-1.9.1+0.aarch64.apple.darwin14/lib/libjulia.1.9.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/local/lib/libjulia.1.9.dylib' (no such file), '/usr/lib/libjulia.1.9.dylib' (no such file)
```
In particular: `(mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))` seems to indicate the Julia install is incompatible with the Python install. **My guess is that your Julia install is ARM (correct) but your Python install is x86_64 (incorrect).**

Just to confirm, you have an M-series macbook, right?

I think your conda environment was set up to use x86_64, rather than using arm64. This not only means that all of your python libraries will run ~2-3x slower (because OSX needs to translate x86_64 -> arm64 using Rosetta), but it also means you can hit library incompatibility issues like the one seen here.

If this guess is correct, I would try to create a new conda environment and make sure to use the correct architecture. This stack overflow answer seems to suggest the right strategy: https://stackoverflow.com/a/65432861/2689923

```bash
conda env config vars set CONDA_SUBDIR=osx-arm64
```

and then re-install all the other libraries in your conda env.

**Reply by MilesCranmer** [reply]
(Another alternative which I don't recommend is to install a x86_64 version of Julia. Then you wouldn't need to create a new conda env. I don't recommend it because it would mean much slower runs.)

**Reply by unary-code** [reply]
"Just to confirm, you have an M-series macbook, right?"
Yeah, I have a M1 Macbook Pro 2020.

"I think your conda environment was set up to use x86_64, rather than using arm64."
Oh, I did not use Conda when I ran "jupyter notebook" on my Macbook terminal.
I do have Conda installed and I have used Conda in the past to run Jupyter Notebook, but I did not use Conda this time.

I think that since I did not use conda (this time) to Jupyter Notebook, that means that my computer must have x86_64 Python.

"If this guess is correct, I would try to create a new conda environment and make sure to use the correct architecture. "
Yeah, that sounds like a good idea: I will just create a new conda environment that has arm64 for Python.

Thank you for explaining!
Thanks also for the stack overflow link!

**Reply by MilesCranmer** [reply]
> Oh, I did not use Conda when I ran "jupyter notebook" on my Macbook terminal.

Jupyter Notebook launched from one Python version can still access other Python versions, and use their libraries (should be an option under "jupyter kernel" in the settings). From the errors you pasted above:

```
  File "/opt/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
```

it's definitely using the anaconda libraries, so maybe you are using the anaconda-installed x86_64 python binary, from inside your jupyter notebook which may have been started from another python.

**Reply by unary-code** [reply]
I still get stuck on the command "!python3 -m pysr install", which still gives me the same error output as I pasted in my earlier comments above.
Even though I followed your advice, specifically, I did "conda env config vars set CONDA_SUBDIR=osx-arm64" while I was in the virtual environment, then I did "conda deactivate pysr_env" (logged out of) the virtual environment then I did "conda activate pysr_env"  (I logged back into the virtual environment), and when I logged in I got a message "WARNING: overwriting environment variables set in the machine
overwriting variable CONDA_SUBDIR" in my terminal, which I think means that the CONDA_SUBDIR variable is osx-arm64 on this environment, while the machine's default value for the CONDA_SUBDIR variable is not osx-arm64.

And then while in the virtual environment on the terminal, I did "jupyter notebook", and then on the Jupyter Notebook interface in the browser in a Python ipynb file I did "!curl -fsSL https://install.julialang.org | sh" followed by "!pip3 install -U pysr" followed by "!python3 -m pysr install", which like I just said, the 
 "!python3 -m pysr install" gave me the same error as above.
Actually, all 3 commands gave me the same output message as I initially pasted in my earlier comments above, which to me I think that means that my running of the command "conda env config vars set CONDA_SUBDIR=osx-arm64" did not really change the virtual environment in a meaningful way.

Perhaps I can double-check. But I think I am using "osx-arm64" now and it is still giving me that error message when I do "!python3 -m pysr install". Any other ideas why? Maybe juliaup is not installed correctly? If not, that's okay, just please let me know either way.

**Reply by MilesCranmer** [reply]
Hm, could you run things directly in terminal (outside of ipynb) to simplify things? It also sounds like it’s not actually switching to the ARM Python version; you may need to delete your conda and re-install the one that matches your architecture for it to respect the osx-arm64 requirement

**Reply by unary-code** [reply]
I just had to uninstall juliaup, and then reinstall juliaup (using "brew install juliaup"), then I had to type "julia" in the terminal in order to download the julia features so "julia" became a recognizable command.
Then I was able to run your PySR package, both 1) in the terminal in my virtual env and 2) from a ipynb file in a Jupyter Notebook in my virtual env!
Thank you for helping me out with this particular issue of not being able to use PySR!

Below is just some extra info (not important):
While I was in the virtual environment, when I did "conda info --system" in the terminal,
The output in the terminal I got was
```
sys.version: 3.8.8 (default, Apr 13 2021, 12:59:45) 
...
sys.prefix: /opt/anaconda3
sys.executable: /opt/anaconda3/bin/python
conda location: /opt/anaconda3/lib/python3.8/site-packages/conda
conda-build: /opt/anaconda3/bin/conda-build
conda-content-trust: /opt/anaconda3/bin/conda-content-trust
conda-convert: /opt/anaconda3/bin/conda-convert
conda-debug: /opt/anaconda3/bin/conda-debug
conda-develop: /opt/anaconda3/bin/conda-develop
conda-env: /opt/anaconda3/bin/conda-env
conda-index: /opt/anaconda3/bin/conda-index
conda-inspect: /opt/anaconda3/bin/conda-inspect
conda-metapackage: /opt/anaconda3/bin/conda-metapackage
conda-render: /opt/anaconda3/bin/conda-render
conda-repo: /opt/anaconda3/bin/conda-repo
conda-server: /opt/anaconda3/bin/conda-server
conda-skeleton: /opt/anaconda3/bin/conda-skeleton
conda-token: /opt/anaconda3/bin/conda-token
conda-verify: /opt/anaconda3/bin/conda-verify
user site dirs: 

CIO_TEST: <not set>
CONDA_DEFAULT_ENV: pysr_env
CONDA_EXE: /opt/anaconda3/bin/conda
CONDA_PREFIX: /opt/anaconda3/envs/pysr_env
CONDA_PREFIX_1: /opt/anaconda3
CONDA_PROMPT_MODIFIER: (pysr_env) 
CONDA_PYTHON_EXE: /opt/anaconda3/bin/python
CONDA_ROOT: /opt/anaconda3
CONDA_SHLVL: 2
CONDA_SUBDIR: osx-arm64
CURL_CA_BUNDLE: <not set>
INFOPATH: /opt/homebrew/share/info:
MANPATH: /opt/homebrew/share/man::
PATH: /opt/anaconda3/bin:/opt/anaconda3/envs/pysr_env/bin:/opt/anaconda3/condabin:/opt/homebrew/bin:/opt/homebrew/sbin:/Library/Frameworks/Python.framework/Versions/3.9/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/opt/X11/bin:/Users/[MY USER]/apache-maven-3.8.7/bin
REQUESTS_CA_BUNDLE: <not set>
SSL_CERT_FILE: <not set>


```

which tells me that the value of CONDA_SUBDIR for this virtual environment was indeed set to osx-arm64 (what I want).

**Reply by MilesCranmer** [reply]
Awesome!!! So what was the fix, just so I can use it to help futures users? Was it the `CONDA_SUBDIR`? Or was it a mix of a few things?

**Reply by unary-code** [reply, marked as answer]
I think it was both CONDA_SUBDIR needed to be osx-arm64, and I needed to type "julia" in my terminal after I had installed juliaup through the "brew install juliaup" command.
Before, I had not typed "julia" after I had installed juliaup through the "brew install juliaup" command, which meant that in the past, my terminal did not recognize the "julia" command.

I had to delete the .julia folder (which for me was located in the top-level directory under /Users/[user-name]), then I had to do "brew install juliaup" then I had to run the command "julia" in the terminal, which installed some things (I'm not sure what). After that, when I typed "julia" in the terminal, the terminal allowed me to type in the Julia coding language (similar to how when you type "python" or "python3" in a terminal, the terminal lets you type in Python coding language). This allowed "!python3 -m pysr install" to work (because before the "!python3 -m pysr install" command had given me the error message that julia was not a recognized command).

I think the reason why I had to re-install juliaup was because my .julia folder for whatever reason only had one juliaup folder which only had one juliaup.json file. When I re-installed juliaup using "brew install juliaup" followed by running the command "julia" in the terminal, after these two commands, I checked the contents of my /Users/[user]/.julia folder, and there were way more files than the last time I had Juliaup.

**Reply by MilesCranmer** [reply]
Thanks for the detailed description!!
